rfcs/proposed/iterators_pass_directly_customization/README.md

```diff
@@ -0,0 +1,151 @@
+pointers refer to data which is device accessible inherently, so no processing is required to pass this type of input
+data to SYCL kernels, we refer to this trait as "passed directly". oneDPL also defines some rules for its provided
+[iterator types](https://uxlfoundation.github.io/oneDPL/parallel_api/iterators.html) to be passed directly to SYCL
+under some circumstances (based usually on their base types).
+
+Internally, these rules are defined with a trait `oneapi::dpl::__ranges::is_passed_directly<T>` which evaluates to
+`std::true_type` or `std::false_type` to indicate whether the iterator type `T` should be passed directly to sycl
+kernels. There exists an unofficial legacy `is_passed_directly` trait which types can define like this:
+`using is_passed_directly = std::true_type;` which is supported within oneDPL. This method is currently used for a
+number of helper types within the SYCLomatic compatibility headers, (`device_pointer`, `device_iterator`,
+`tagged_pointer`, `constant_iterator`, `iterator_adaptor`). There is no official public API for users who want to
+create their own iterator types which could be passed directly to SYCL kernels, this is a gap that should be filled 
+with an official public API.
+
+Without something like this users are forced to only rely upon our provided iterator types, or reach into implementation
+details which are not part of oneDPL's specified interface.
+
+## Proposal
+
+Create a customization point `oneapi::dpl::is_passed_directly_in_onedpl_device_policies` free function which allows
+users to mark their types as passed directly:
+
+```
+template <typename T>
+constexpr bool is_passed_directly_in_onedpl_device_policies(const T&);
+```
+
+oneDPL will provide a default implementation which will defer to the existing trait:
+
+```
+template <typename T>
+constexpr
+bool
+is_passed_directly_in_onedpl_device_policies(const T&)
+{
+	return oneapi::dpl::__ranges::is_passed_directly_v<T>;
+}
+```
+
+Below is a simple example of a type and customization point definition which is always passed directly.
+
+```
+namespace user
+{
+
+    struct my_passed_directly_type
+    {
+        /* unspecified user definition */
+    };
+
+    constexpr
+    bool
+    is_passed_directly_in_onedpl_device_policies(const my_passed_directly_type&)
+    {
+        return true;
+    }
+} //namespace user
+```
+
+Users can use any constexpr logic based on their type to determine if the type can be passed directly into a SYCL kernel
+without any processing. Below is an example of a type which contains a pair of iterators, and should be treated as
+passed directly if and only if both base iterators are also passed directly. oneDPL will use this customization point
+internally when determining how to handle incoming data, picking up any user defined customizations.
+
+When using device policies, oneDPL will check iterator types by calling `is_passed_directly_in_onedpl_device_policies`
```

Comment:
Is it supposed to be a compile-time evaluation or a run-time function call?

Reply:
As I experiment a little more, I'm realizing that perhaps we can't combine the customization point with a compile-time evaluation requirement, or some other requirements on the iterator types themselves like default constructible.

I'm running into issues attempting to use it within `enable_if` / SFINAE without some way to take the type alone and turn it into a const lvalue ref argument for ADL to work.  If I can call a default constructor, it works, but I'm not sure what other option there is.  

At first I thought `std::declval` was the answer, but it is not.  It is, in fact, both not `constexpr`, and it would be being used in an evaluated expression here. I'll continue experimenting...

Reply:
I think its only a roadblock for using it in SFINAE, and not in general for `constexpr`. I believe we should be able to refactor our code to lean instead on `if constexr` for "passed directly" checks.

Reply:
I found a way to deal use it with SFINAE as well, it required returning `std::true_type{}` or `std::false_type` rather than `bool`.  This seems like the right way to go. I've updated the PR accordingly.